synapse/app/homeserver.py

```diff
@@ -361,9 +357,43 @@ def setup(config_options):
     logger.info("Database prepared in %s.", config.database_config['name'])
 
     hs.setup()
-    hs.start_listening()
 
+    @defer.inlineCallbacks
     def start():
+        # Check if the certificate is still valid.
+        cert_days_remaining = hs.config.is_disk_cert_valid()
+
+        if hs.config.acme_enabled:
+            # If ACME is enabled, we might need to provision a certificate
+            # before starting.
+            acme = hs.get_acme_handler()
+
+            # Start up the webservices which we will respond to ACME challenges
+            # with.
+            try:
+                yield acme.start_listening()
+            except defer.FirstError as e:
```

Comment:
it feels fragile to tie this to FirstError. should it not extend to any Exception? (and indeed, should it not apply to the whole of `start` ?)

Reply:
Yeahhhh I guess catching the entirety (and then special casing FirstError, which can only come from the DeferredList to provide a nicer log message) will be better.

Not sure about making the entirety of `start` in a block. I guess we ought to, since there's now more in it?

Reply:
yeah I guess we ought to, really. Mostly it felt odd to add a special-case for this one line.

Reply:
done